pkg/vpcmodel/groupingSelfLoop.go

```diff
@@ -0,0 +1,333 @@
+	"strings"
+)
+
+// extends grouping by considering self loops as don't care https://github.com/np-guard/vpc-network-config-analyzer/issues/98
+// e.g. a => b,c   b => a, c   and   c => a,b   is actually a clique a,b,c => a,b,c
+// a => b,c, b => c can be presented in one line as  a,b => b,c
+
+// After the basic grouping, which is of worst time complexity O(n^2), we optimize grouping treating self loops as don't care.
+// Intuitively, we check if two GroupedConnLine can be merged treating self loops as don't care
+// 1. groupsToBeMerged find couples of GroupedConnLine that should be merged using the alg below
+// mergeSelfLoops merges the groupsToBeMerged:
+// 2. It creates sets of GroupedConnLine s.t. each set should be merged
+//			note that the "should be merged" is an equivalence relation
+// 3. It merges them
+
+// alg: GroupedConnLine to be merged:
+// GroupedConnLine whose distance is an empty set should be merged;
+// the claim below guarantees that the result is coherent
+//
+// The distance between two GroupedConnLine is defined as following:
+// Let l_1 be a line with source s_1 and dest d_1 and let l_2 be a line with source s_2 and dest d_2.
+// l_1 / l_2 is the vsis/subnets in d_1 that are not in d_2 minus the single vsi/subnet in s_1 if |s_1| = 1
+//
+// The distance between lines l_1 and l_2 is l_1 / l_2 union l_2 / l_1
+//
+// claim: if the distance between line l_1 and l_2 is an empty set and
+//		 the distance between lines l_2 and l_3 is an empty set
+//		 then so is the distance between l_1 and l_3
+
+func (g *GroupConnLines) extendGroupingSelfLoops(groupingSrcOrDst map[string][]*GroupedConnLine,
+	srcGrouping bool) map[string][]*GroupedConnLine {
+	toMergeCouples := g.groupsToBeMerged(groupingSrcOrDst, srcGrouping)
+	return mergeSelfLoops(toMergeCouples, groupingSrcOrDst, srcGrouping)
+}
+
+func (g *GroupConnLines) groupsToBeMerged(groupingSrcOrDst map[string][]*GroupedConnLine, srcGrouping bool) (toMergeCouples [][2]string) {
+	toMergeCouples = make([][2]string, 0)
+	// the to be grouped src/dst in set representation, will be needed to compute potential groups to be merged
+	// and to compute the deltas
+	setsToGroup := createGroupingSets(groupingSrcOrDst, srcGrouping)
+	relevantKeys := relevantKeysToCompare(groupingSrcOrDst)
+	keyToMergeCandidates := g.findMergeCandidates(groupingSrcOrDst, srcGrouping, setsToGroup, relevantKeys)
+
+	for _, key := range relevantKeys {
+		keyLines := groupingSrcOrDst[key]
+		//  is there a different line s.t. the keyLines were not merged only due to self loops?
+		// 	going over all couples of items: merging them if they differ only in self loop element
+		// findMergeCandidates of a singleton 'key' are all lines in which the group contains 'key'
+		//  if key is not a singleton then findMergeCandidates will be empty
+		mergeCandidates, ok := keyToMergeCandidates[key]
+		if !ok {
+			continue
+		}
+		for candidate := range mergeCandidates {
+			candidateLines := groupingSrcOrDst[candidate]
```

Comment:
Now this will be the case only if both `x` and `y` are singeltons. Not sure its worth the trouble in this case. In fact, even complexity wise its not worth while. Its no longer a matrix.